### ISSUE TITLE
Dev build crashes on macOS

### ISSUE THREAD

**Issue by FormalSnake** [conaticus/FileExplorer#26]
cargo tauri dev                 
                                                                                                                                                                                                                                         
  Running BeforeDevCommand (`yarn dev`)
yarn run v1.22.19
warning ../../package.json: No license field
$ vite

  VITE v4.3.9  ready in 408 ms

  ➜  Local:   http://127.0.0.1:1420/
  ➜  Network: use --host to expose
        Info Watching /Users/kyandesutter/development/FileExplorer/src-tauri for changes...
   Compiling file-explorer v0.0.0 (/Users/kyandesutter/development/FileExplorer/src-tauri)
    Finished dev [unoptimized + debuginfo] target(s) in 4.08s
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Error("EOF while parsing a value", line: 1, column: 0)', src/filesystem.rs:176:63
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
fatal runtime error: failed to initiate panic, error 5

Note from @conaticus for developers:
I have implemented a mechanism to stop the program from crashing if it is unable to deserialize the cache, but it leads to the program recaching every time it is re-run. Look out for a log message to see if the deserialization failed. There is also a panic on this line in the `search.rs` when searching on MacOS:

![image](https://github.com/conaticus/FileExplorer/assets/72096531/c3c86fe0-4748-40ae-bb27-e960147e86aa)

This is likely a string mismatch causing the program to fail when retrieving the volume from the cache.

**Comment by FormalSnake** [comment]
The problem with that is the window becomes blank

**Comment by FormalSnake** [comment]
and unresponsive.

**Comment by FormalSnake** [comment]
I have let it run for an hour now and it still isn't done, and didn't write anything.

**Comment by RaphGL** [comment]
it should only take a couple minutes at best. something must be going wrong here. I'll see what we can do, I don't have a mac tho so I'll have to relying on other people's help.

**Comment by HenryRoutson** [comment]
> This is likely a string mismatch causing the program to fail when retrieving the volume from the cache.

mount_pnt is actually an empty string "" on mac os, not sure why it would not be in the cache
Something about windows and unix having different file systems

**Comment by RaphGL** [comment]
> > This is likely a string mismatch causing the program to fail when retrieving the volume from the cache.
> 
> mount_pnt is actually an empty string "" on mac os, not sure why it would not be in the cache Something about windows and unix having different file systems

I'm on linux so it's definitely not a unix thing as it works for me and someone else on linux has tried it and it works as well. We're using a [cross platform crate](https://docs.rs/sysinfo/latest/sysinfo/index.html) to retrieve mountpoints which are then stored in the cache. I'm not exactly sure whether this is an error on our part or if there's a bug on the crate itself.

**Comment by jokorone** [comment]
it seems to me the current code is trying to call `.unwrap()` on a [poisened mutex](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning) when accessing state in the `run_cache_interval` loop.

```
        loop {
            interval.tick().await;

            let mut guard = match state_clone.lock() {
                Ok(state) => state,
                Err(poison_error) => poison_error.into_inner(),
            };

            save_to_cache(&mut guard);
        }
```
did the trick for.

**Comment by HenryRoutson** [comment]
> it seems to me the current code is trying to call `.unwrap()` on a [poisened mutex

hmmm fix still fails on my machine (M1 Pro Ventura), 
Might be a different problem.

Why would a poisoned mutex be created? What thread is calling panic! and not taking down the program?

**Comment by jokorone** [comment]
@HenryRoutson at first I experienced the same, but deleting the `system_cache.json` and resolved the issue.

From the [docs](https://doc.rust-lang.org/std/sync/struct.Mutex.html#poisoning):
> [...] where a mutex is considered poisoned whenever a thread panics while holding the mutex. Once a mutex is poisoned, all other threads are unable to access the data by default as it is likely tainted (some invariant is not being upheld).

**Comment by RaphGL** [comment]
I've tested it, it works for me as well. 
I think the issue is that the code keeps triggering the recaching without checking if the previous caching routine is done or not. If the caching takes more than 30 seconds it will spawn a new routine that will try to lock the mutex and panic causing the poisoning.

@conaticus to fix this we could use a channel to decide when to restart the caching if necessary.

**Comment by HenryRoutson** [comment]
Also formal snake double check you have run yarn and installed tauri-cli,
i did just cargo run out of intuition first time and got a blank screen

cargo install tauri-cli
yarn

**Comment by FormalSnake** [comment]
Where should I put the provided code snippet?

**Comment by HenryRoutson** [comment]
paste this in your terminal 

cargo install tauri-cli
yarn

and then run 

cargo tauri dev

**Comment by FormalSnake** [comment]
I know that, but I mean where do I put this:

``        loop {
            interval.tick().await;

            let mut guard = match state_clone.lock() {
                Ok(state) => state,
                Err(poison_error) => poison_error.into_inner(),
            };

            save_to_cache(&mut guard);
        }``

**Comment by HenryRoutson** [comment]
So search for interval.tick and you should find a loop in cache.rs you want to replace

**Comment by FormalSnake** [comment]
I am not sure but i do not think it is working

**Comment by FormalSnake** [comment]
I deleted the cache and i am running it now.

**Comment by FormalSnake** [comment]
it's been running for 9 minutes and it didn't add anything to the system_cache, and it is still unresponsive

**Comment by jokorone** [comment]
@FormalSnake how much diskspace are you using? 👀

**Comment by RaphGL** [comment]
> it's been running for 9 minutes and it didn't add anything to the system_cache, and it is still unresponsive

there's no system cache file anymore, at least not on the repo. cache now goes to the appropriate location depends on the operating system. check [this](https://docs.rs/dirs/5.0.1/dirs/fn.cache_dir.html) to know where it goes on your system.

**Comment by FormalSnake** [comment]
it did create a new json for me tho. Is there a new commit?

**Comment by jokorone** [comment]
> I've tested it, it works for me as well. I think the issue is that the code keeps triggering the recaching without checking if the previous caching routine is done or not. If the caching takes more than 30 seconds it will spawn a new routine that will try to lock the mutex and panic causing the poisoning.
> 
> @conaticus to fix this we could use a channel to decide when to restart the caching if necessary.

Only reading this now.

There's no need to re-cache if another spawn is already doing just that. Accessing the poisoned mutex would create issues in the long run I suspect.

As with the example of @FormalSnake with a loaded Disk (and a fast chipset!, Imagine a movie buff with TBs and a Intel from 2012), I guess 30sec isn't enough and thus it kinda makes my PR obsolete 🥲

**Comment by FormalSnake** [comment]
so is mine supposed to take a bit?

**Comment by RaphGL** [comment]
@FormalSnake try out 
```
git pull origin/dev
```
the og dev branch has kind of gotten obliterated. I think you're running a few days old commit. Try running that command or just recloning the project (the lazy route).

**Comment by conaticus** [comment]
> it did create a new json for me tho. Is there a new commit?

If you're on the latest version it should cache to your system's cache directory. If it fails to read the cache file it will print a log message in the console.

As mentioned in the edit there are some known issues with MacOS that need to be addressed. We are waiting on a mac developer familiar with Rust to assist with this.

@RaphGL's idea is a valid problem that a new issue should be created for however I am not convinced this is neccesarily the problem @FormalSnake is experiencing.

@FormalSnake please could you pull the latest version on the **dev** branch and let me know the following:
- Any panics that occur
- If there is a log message ever printed saying `Failed to deserialize the cache from disk, recaching...`
- Whether there is a file named `file-explorer.cache.bin` in your `/Users/{Your Name}/Library/Caches`
- The file size of the `file-explorer.cache.bin` if it exists

Thanks

**Comment by jokorone** [comment]
What exactly is the idea with updating the cache in an interval?

Seems to me this eats up some memory? Not sure if neglectible.
A watcher would do the same thing, but only run when something actually changed.

**Comment by FormalSnake** [comment]
It loaded really quickly now! It opened to a Loadin... screen and it created a bin file in my cache folder

**Comment by conaticus** [comment]
> What exactly is the idea with updating the cache in an interval?
> 
> Seems to me this eats up some memory? Not sure if neglectible. A watcher would do the same thing, but only run when something actually changed.

The cache is stored on disk and this is a very big file that takes a few seconds to update. Filesystem changes actually occur in the background very frequently and this would invoke the program to write the file too often. It's not a perfect solution but a temporary one. I am about to open an issue about the 30s delay edge case.

**Comment by conaticus** [comment]
> It loaded really quickly now! It opened to a Loadin... screen and it created a bin file in my cache folder

Great! Any errors or crashes? Does the searching work okay?

**Comment by FormalSnake** [comment]
it sstill loading currently

**Comment by FormalSnake** [comment]
I am going away currently so I will say if it loaded or not.

**Comment by conaticus** [comment]
> I am going away currently so I will say if it loaded or not.

If it has taken this long then @RaphGL is correct. Sounds like it's trying to recache before finishing the first cache resulting in an infinite loop.

**Comment by FormalSnake** [comment]
It is still going

**Comment by conaticus** [comment]
Even on slower systems this shouldn't take longer than 5 minutes. As mentioned it will likely be an infinite loop happening here. Thanks for your help and I apologise that this isn't working for you currently.

**Comment by FormalSnake** [comment]
No need to apologize. I hope we figure it out

**Comment by FormalSnake** [comment]
After a few hours it did manage to cache!

**Comment by FormalSnake** [comment]
searching is bad tho:

thread 'tokio-runtime-worker' panicked at 'called `Option::unwrap()` on a `None` value', src/search.rs:85:59
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/filesystem/cache.rs:139:49
thread 'notify-rs fsevents loop' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/filesystem/cache.rs:48:48
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: PoisonError { .. }', src/search.rs:82:34

**Comment by FormalSnake** [comment]
And restarting the app makes it go back to Loading...

**Comment by RaphGL** [comment]
@FormalSnake if you start searching before entering a disk/volume you'll get a panic, is that what u did?

**Comment by FormalSnake** [comment]
it also doesnt work inside a volume

**Comment by RaphGL** [comment]
We should probably wait for the upcoming work on the search stuff before checking that I think. 
I think this issue can now be closed as you've already managed to build and run the program.
Please create an issue if you get this error after the search refactor/rewrite is merged.

**Comment by FormalSnake** [comment]
No problem! Thanks for all the support!

**Comment by conaticus** [comment]
These still need addressing in future
